### PR TITLE
msg: Return a handle instead of kernel pointer in msgRecv

### PR DESCRIPTION
**Pull request by agkaminski**
Breaking API change!

Add idtree from libphoenix, some MISRA improvements

JIRA: RTOS-606

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [x] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (ia32-generic-qemu).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work (a whole lot).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: lib/idtree.c [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+	}
+}
+
+
+static int lib_idtreeGapcmp(rbnode_t *n1, rbnode_t *n2)
+{
+	idnode_t *r1 = lib_treeof(idnode_t, linkage, n1);
+	idnode_t *r2 = lib_treeof(idnode_t, linkage, n2);
+	rbnode_t *child = NULL;
+	int ret = 1;
```
Comment (badochov) [inline lib/idtree.c:41]: IMHO `ret` should be set to zero as that's its de facto default value, or not set at all as whenevere child != NULL then ret is set

File: lib/idtree.c [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+	if (r1->lmaxgap > 0 && r1->rmaxgap > 0) {
+		if (r2->id > r1->id) {
```
Comment (badochov) [inline lib/idtree.c:44]: r2-> id is always zero so there's no need for this branch thus this whole if else can be simplified to 
```c
if (r1->lmaxgap > 0) {
		child = n1->left;
		ret = 1;
	}
	else if (r1->rmaxgap > 0) {
		child = n1->right;
		ret = -1;
	}
```

Comment (agkaminski) [reply lib/idtree.c:44]: Why do you think it's always zero? It's not obligatory to have id == 0 element in the tree:
- allocate id = 0,
- allocate id = 1,
- remove id = 0
and we have "id = 0"-less tree

File: lib/idtree.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * id allocating tree
+ *
+ * Copyright 2018 Phoenix Systems
+ * Author: Jan Sikorski
```
[the remark this answers is not in the document]
Comment (agkaminski) [reply lib/idtree.h]: It is, I've imported it from libphoenix. I modified it quite substantially now though, so I've added myself as a coauthor

File: proc/msg-nommu.c [hunk 4]
```diff
@@ -55,11 +57,20 @@ int proc_send(u32 port, msg_t *msg)
...
+
+			if ((err != EOK) && (kmsg.state == msg_waiting)) {
+				LIST_REMOVE(&p->kmessages, &kmsg);
+				break;
+			}
+		}
+
+		if (kmsg.state == msg_responded) {
+			err = EOK; /* Don't report EINTR if we got the response already */
+		}
```
[the remark this answers is not in the document]
Comment (agkaminski) [reply proc/msg-nommu.c]: I guess it's better to do this in one place, so I've modified it as suggested

File: lib/rb.h [hunk 5]
```diff
@@ -19,10 +19,10 @@
 //#include "../hal/hal.h"
```
[the remark this answers is not in the document]
Comment (agkaminski) [reply lib/rb.h]: Not really in scope of this PR, but ok

File: lib/idtree.c (hunk 1 above)
Comment (agkaminski) [reply lib/idtree.c]: As it turns out we don't need a child at all

Comment (agkaminski) [reply lib/idtree.c]: Retreat, it's need afterall

File: lib/idtree.c (hunk 2 above)
Comment (badochov) [reply lib/idtree.c:44]: this function is only used in `lib_rbFindEx` where n2 is set to `n` whose id is set to 0. During the execution of `lib_rbFindEx` the second argument is constant thus r2->id is always 0

Comment (agkaminski) [reply lib/idtree.c:44]: I'd rather have this code more generic than it's necessary right now, so I won't get bitten in the future

Comment (badochov) [reply lib/idtree.c:44]: Ok

**Review by badochov — APPROVED** [review]

**Review by anglov — APPROVED** [review]
Tested by hand within DEND project and works fine

**Pull request merged**